optimum/exporters/tasks.py

```diff
@@ -1624,10 +1668,13 @@ def infer_library_from_model(
                     library_name = "timm"
                 elif hasattr(model_config, "_diffusers_version"):
                     library_name = "diffusers"
-                elif any(file_path.startswith("sentence_") for file_path in all_files):
-                    library_name = "sentence_transformers"
-                else:
```

Comment:
can we add back the case of transformers? When testing the inference, it seems that library could be None in this case whereas it should be transformers:
```
============================================================ FAILURES ============================================================
__________________ NeuronModelForFeatureExtractionIntegrationTest.test_compare_to_transformers_dyn_bs_0_albert ___________________

a = (<tests.inference.test_modeling.NeuronModelForFeatureExtractionIntegrationTest testMethod=test_compare_to_transformers_dyn_bs_0_albert>,)
kw = {}

    @wraps(func)
    def standalone_func(*a, **kw):
>       return func(*(a + p.args), **p.kwargs, **kw)

../pyvenv/aws_neuron_venv_2.16/lib/python3.8/site-packages/parameterized/parameterized.py:620: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/inference/test_modeling.py:197: in test_compare_to_transformers_dyn_bs
    neuron_model_dyn = NeuronModelForFeatureExtraction.from_pretrained(
../pyvenv/aws_neuron_venv_2.16/lib/python3.8/site-packages/optimum/modeling_base.py:349: in from_pretrained
    library_name = TasksManager.infer_library_from_model(model_id, subfolder, revision, cache_dir)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'optimum.exporters.tasks.TasksManager'>, model_name_or_path = '/tmp/albert_dyn_bs_true_feature-extraction_r02o_ps_'
subfolder = '', revision = None, cache_dir = None, library_name = None

    @classmethod
    def infer_library_from_model(
        cls,
        model_name_or_path: Union[str, Path],
        subfolder: str = "",
        revision: Optional[str] = None,
        cache_dir: str = huggingface_hub.constants.HUGGINGFACE_HUB_CACHE,
        library_name: Optional[str] = None,
    ):
        """
        Infers the library from the model repo.
    
        Args:
            model_name_or_path (`str`):
                The model to infer the task from. This can either be the name of a repo on the HuggingFace Hub, an
                instance of a model, or a model class.
            subfolder (`str`, defaults to `""`):
                In case the model files are located inside a subfolder of the model directory / repo on the Hugging
                Face Hub, you can specify the subfolder name here.
            revision (`Optional[str]`, *optional*, defaults to `None`):
                Revision is the specific model version to use. It can be a branch name, a tag name, or a commit id.
            cache_dir (`Optional[str]`, *optional*):
                Path to a directory in which a downloaded pretrained model weights have been cached if the standard cache should not be used.
            library_name (`Optional[str]`, *optional*):
                The library name of the model. Can be any of "transformers", "timm", "diffusers", "sentence_transformers".
        Returns:
            `str`: The library name automatically detected from the model repo.
        """
        if library_name is not None:
            return library_name
    
        full_model_path = Path(model_name_or_path) / subfolder
    
        if not full_model_path.is_dir():
            model_info = huggingface_hub.model_info(model_name_or_path, revision=revision)
            library_name = getattr(model_info, "library_name", None)
    
            # sentence-transformers package name is sentence_transformers
            if library_name is not None:
                library_name = library_name.replace("-", "_")
    
        if library_name is None:
            all_files, _ = TasksManager.get_model_files(model_name_or_path, subfolder, cache_dir)
    
            if "model_index.json" in all_files:
                library_name = "diffusers"
            elif CONFIG_NAME in all_files:
                model_config = PretrainedConfig.from_pretrained(
                    model_name_or_path, subfolder=subfolder, revision=revision
                )
    
                if hasattr(model_config, "pretrained_cfg") or hasattr(model_config, "architecture"):
                    library_name = "timm"
                elif hasattr(model_config, "_diffusers_version"):
                    library_name = "diffusers"
            elif (
                any(file_path.startswith("sentence_") for file_path in all_files)
                or "config_sentence_transformers.json" in all_files
            ):
                library_name = "sentence_transformers"
            else:
                library_name = "transformers"
    
        if library_name is None:
>           raise ValueError(
                "The library name could not be automatically inferred. If using the command-line, please provide the argument --library {transformers,diffusers,timm,sentence_transformers}. Example: `--library diffusers`."
            )
E           ValueError: The library name could not be automatically inferred. If using the command-line, please provide the argument --library {transformers,diffusers,timm,sentence_transformers}. Example: `--library diffusers`.

../pyvenv/aws_neuron_venv_2.16/lib/python3.8/site-packages/optimum/exporters/tasks.py:1680: ValueError
```

Reply:
For which model is that? I don't see how this could occur.

Reply:
It's `hf-internal-testing/tiny-random-AlbertModel` and all inference tests of encoder models fail.

```
(Pdb) all_files
['special_tokens_map.json', 'tokenizer.json', 'config.json', 'tokenizer_config.json', 'spiece.model', 'model.neuron']
(Pdb) CONFIG_NAME in all_files
True
(Pdb) model_config
PretrainedConfig {
  "_name_or_path": "hf-internal-testing/tiny-random-AlbertModel",
  "architectures": [
    "AlbertModel"
  ],
  "attention_probs_dropout_prob": 0.1,
  "bos_token_id": 2,
  "classifier_dropout_prob": 0.1,
  "embedding_size": 128,
  "eos_token_id": 3,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 36,
  "initializer_range": 0.02,
  "inner_group_num": 1,
  "intermediate_size": 37,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "neuron": {
    "auto_cast": null,
    "auto_cast_type": null,
    "compiler_type": "neuronx-cc",
    "compiler_version": "2.12.54.0+f631c2365",
    "disable_fallback": false,
    "disable_fast_relayout": false,
    "dynamic_batch_size": true,
    "input_names": [
      "input_ids",
      "attention_mask",
      "token_type_ids"
    ],
    "model_type": "albert",
    "optlevel": "2",
    "output_attentions": false,
    "output_hidden_states": false,
    "output_names": [
      "last_hidden_state",
      "pooler_output"
    ],
    "static_batch_size": 1,
    "static_sequence_length": 32
  },
  "num_attention_heads": 6,
  "num_hidden_groups": 6,
  "num_hidden_layers": 6,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "task": "feature-extraction",
  "torch_dtype": "float32",
  "transformers_version": "4.36.2",
  "type_vocab_size": 16,
  "vocab_size": 30000
}
```

Reply:
Well maybe it's for neuron, not sure how onnx files are organised. Maybe we should enable passing lib name to from_pretrained as well?

Reply:
```python
from optimum.exporters import TasksManager

library = TasksManager.infer_library_from_model("hf-internal-testing/tiny-random-AlbertModel")
print(library)
```

works as expected. Could you provide a reproduction?

Reply:
Thank you @JingyaHuang, this should be fixed, could you give a try?

Reply:
Yay it works! Thxxxxx!